apps/chromealive-core/injected-scripts/generateSelectorMap.ts

```diff
@@ -0,0 +1,363 @@
+import type {
+  ISelectorMap,
+  ISelectorOption,
+} from '@ulixee/apps-chromealive-interfaces/ISelectorMap';
+
+const QuerySortOrder = {
+  tag: 1,
+  id: 2,
+  class: 3,
+  attr: 4,
+  index: 5,
+};
+
+interface ITarget {
+  element: Element;
+  heroNodeId: number;
+  selectorOptions: ISelectorOption[];
+}
+
+type IAncestors = ITarget[];
+type Bitmask = bigint;
+
+interface ILayerOption {
+  element: Element;
+  selectorOption: ISelectorOption;
+  index: number;
+}
+
+const RankByType = {
+  id: 1,
+  class: 2,
+  tag: 3,
+  attr: 4,
+  index: 2,
+};
+
+export interface ISelectorMapContext {
+  target: ITarget;
+  ancestors: IAncestors;
+  boundary: ParentNode;
+  selectors: {
+    targetBitmask: Bitmask;
+    ancestorBitmasks: Bitmask[]; // [body, div, ... ] - not including target
+    rank: number;
+    selector: string;
+  }[];
+}
+
+/**
+ * Ancestor patterns are just bitmasks of the Permutations 2^<ancestor length> -1. 2 = ['11','01','10'] = [3,1,2]
+ * Selector options are also just bitmasks of available options 2^[options.length] -1.
+ * Start at current element - check for any unique patterns.
+ * Count querySelectorAll to see if there's exactly 1 matching element
+ */
+
+export default function generateSelectorMap(
+  element: Element,
+  boundary: ParentNode = document,
+  iterationsBeforeFirstBreak = 50e3,
+): ISelectorMapContext {
+  const target = generateTarget(element, boundary);
+  const ancestors = generateAncestors(element, boundary);
+  const context: ISelectorMapContext = {
+    target,
+    ancestors,
+    boundary,
+    selectors: [],
+  };
+
+  const emptyAncestorMasks = ancestors.map(() => 0n);
+  getSelectorCounts(context, emptyAncestorMasks);
+
+  // sorted by least layers involved
+  const allSelectorOptions = generateAllLayerOptions(ancestors);
+  const maxSelectorCombinationBitmask = getMaxBitmaskValue(allSelectorOptions);
+
+  // console.log('Checking combinations. First 50...', allSelectorOptions);
+  // get bitmask of all the combined options (NOTE: we could just combine here, but just means creating arrays upfront)
+
+  const selectorHistory = new Set<string>();
+  let counter = 0n;
+  let nextBreakpoint = iterationsBeforeFirstBreak;
+  iterateBitmaskBySubsets(maxSelectorCombinationBitmask, mask => {
+    const selectorPermutations = decodeBitmask(mask, allSelectorOptions);
+    const { selector, rankSum, layerBitmasks } = createAncestorSelector(
+      selectorPermutations,
+      context.ancestors,
+    );
+    if (selectorHistory.has(selector))
+      console.warn('Selector already tried', selector, {
+        rankSum,
+        layerBitmasks,
+        selectorPermutations,
+      });
+    counter += getSelectorCounts(context, layerBitmasks, selector, rankSum);
+
+    if (counter > nextBreakpoint) {
+      nextBreakpoint += 10e3;
+      console.log('Checking for matches after %s permutations', counter, context.selectors.length);
```

Comment:
Should this console.log be here?

Reply:
I can remove it, but I was finding it useful to keep an eye on.

Reply:
Ok. We can keep it in for now... at some point we should probably push this to the UI.